crates/sparrow-runtime/src/execute/output/object_store.rs

```diff
@@ -1,209 +1,263 @@
+            .into_report()
+            .change_context(Error::Upload)?;
+
+        match file_type {
+            FileType::Unspecified => todo!(),
+            FileType::Parquet => {
+                let writer = parquet::arrow::AsyncArrowWriter::try_new(
+                    writer,
+                    schema,
+                    UPLOAD_BUFFER_SIZE_IN_BYTES,
+                    None,
+                )
+                .into_report()
+                .change_context(Error::Write)?;
+                Ok(Self::Parquet { url, writer })
+            }
+            FileType::Csv => {
+                // Write an empty batch so headers are written.
+                let mut buffer = Vec::new();
+                let mut csv_writer = arrow::csv::WriterBuilder::new()
+                    .has_headers(true)
+                    .build(&mut buffer);
+                csv_writer
+                    .write(&RecordBatch::new_empty(schema))
+                    .into_report()
+                    .change_context(Error::Write)?;
+                std::mem::drop(csv_writer);
+                writer
+                    .write_all(&buffer)
+                    .await
+                    .into_report()
+                    .change_context(Error::Write)?;
+
+                // Then return the state
+                Ok(Self::Csv { url, writer })
+            }
+        }
+    }
+
+    async fn write(&mut self, batch: RecordBatch) -> error_stack::Result<(), Error> {
+        match self {
+            WriterState::Parquet { writer, .. } => {
+                writer
+                    .write(&batch)
+                    .await
+                    .into_report()
+                    .change_context(Error::Write)?;
+                Ok(())
+            }
+            WriterState::Csv { writer, .. } => {
+                // Re-allocating the buffer on every CSV write is likely inefficient.
+                // But, we anticipate CSV output being used with small / example cases so
+                // we accept this for now. Ideally, we could have an async CSV writer that
+                // allows us to flush to the object store rather thna growing the buffer.
+                // In the absence of that, we should support growing the buffer, but also
+                // try to slice the batch into small enough chunks that we don't need to
+                // and upload in a sequence of "pages".
+                let mut buffer = Vec::new();
+                let mut csv_writer = arrow::csv::WriterBuilder::new()
+                    .has_headers(false)
+                    .build(&mut buffer);
+                csv_writer
+                    .write(&batch)
+                    .into_report()
+                    .change_context(Error::Write)?;
+                std::mem::drop(csv_writer);
+                writer
+                    .write_all(&buffer)
+                    .await
+                    .into_report()
+                    .change_context(Error::Write)?;
```

Comment:
What changed in this implementation -- why do we need to allocate a buffer for each batch? Due to the `writer` being a trait now?

Reply:
Because we're using an `AsyncWrite` which requires we `await`. Unfortunately, the CSV writer only supports the synchronous `Write` trait. So we can either:

1. Run the future on the current thread, blocking until it completes.
2. Run synchronous code to write to a buffer, then use async code to write the buffer to the object store.

Given that CSV is likely to be used for smaller results, it seems like (2) is better, since it allows the thread to do other things while waiting (although, if we had more threads than cores, it may be OK because the thread would likely sleep while waiting). This seems to be the approach that Arrow suggests for CSV (see https://github.com/apache/arrow-rs/issues/3740).